demos/mqtt/mqtt_demo_connection_sharing/mqtt_demo_connection_sharing.c

```diff
@@ -0,0 +1,1601 @@
+        }
+
+        /* The humidity callback should have been invoked for handling the incoming
+         * PUBLISH message on the humidity topic. */
+        else if( globalReceivedHumidityData != true )
+        {
+            returnStatus = EXIT_FAILURE;
+            LogError( ( "Registered callback was not invoked on sending a PUBLISH to "
+                        "the humidity topic: Topic=%s", DEMO_HUMIDITY_TOPIC ) );
+        }
+    }
+
+    /* Subscribe to the precipitation topic filter, this time also without any wildcard characters,
+     * so that we can receive incoming PUBLISH message only on the same topic from the broker. */
+    if( returnStatus == EXIT_SUCCESS )
+    {
+        /* Register subscription callback for the wildcard notification topic filter. */
+        ( void ) SubscriptionManager_RegisterCallback( DEMO_PRECIPITATION_TOPIC,
+                                                       DEMO_PRECIPITATION_TOPIC_LENGTH,
+                                                       precipitationDataCallback );
+
+        /* The client is now connected to the broker. Subscribe to the precipitation topic
+         * by sending a subscribe packet. This client will then publish to the same topic it
+         * subscribed to, so it will expect all the messages it sends to the broker
+         * to be sent back to it from the broker. This demo uses QOS2 in Subscribe,
+         * therefore, the Publish messages received from the broker will have QOS2. */
+        LogInfo( ( "Subscribing to the MQTT topic %s.",
+                   DEMO_PRECIPITATION_TOPIC ) );
+        returnStatus = subscribeToTopic( &mqttContext,
+                                         DEMO_PRECIPITATION_TOPIC,
+                                         DEMO_PRECIPITATION_TOPIC_LENGTH );
+    }
+
+    if( returnStatus == EXIT_SUCCESS )
+    {
+        /* Process incoming packet from the broker. Acknowledgment for subscription
+         * ( SUBACK ) will be received here. However after sending the subscribe, the
+         * client may receive a publish before it receives a subscribe ack. Since this
+         * demo is subscribing to the topic to which no one is publishing, probability
+         * of receiving publish message before subscribe ack is zero; but application
+         * must be ready to receive any packet. This demo uses MQTT_ProcessLoop to
+         * receive packet from network. */
+        mqttStatus = MQTT_ProcessLoop( &mqttContext, MQTT_PROCESS_LOOP_TIMEOUT_MS );
+
+        if( mqttStatus != MQTTSuccess )
+        {
+            returnStatus = EXIT_FAILURE;
+            LogError( ( "MQTT_ProcessLoop returned with status = %u.",
+                        mqttStatus ) );
+
+            /* Remove registered callback for the notify wildcard topic filter. */
+            ( void ) SubscriptionManager_RemoveCallback( DEMO_PRECIPITATION_TOPIC,
+                                                         strlen( DEMO_PRECIPITATION_TOPIC ) );
+        }
+    }
```

Comment:
This Register callback -> subscribe -> process loop -> remove callback on failure pattern is repetitive and could very easily be a separate function.

Reply:
Updated the `subscribeToTopic` function to contain the call to `ProcessLoop` so that it represents a complete _subscribe_ operation.

Also, created a separate utility, `subscribeToAndRegisterTopicFilter`, for containing the _subscribe_ and _registration_ logic